src/lfortran/semantics/ast_body_visitor.cpp

```diff
@@ -2226,6 +2226,61 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         current_scope = parent_scope;
     }
 
+    /**
+     * Checks compatibility between target and value arrays for assignment operations.
+     *
+     * This function verifies that the target and value arrays are compatible for assignment by ensuring
+     * - they have the same rank and
+     * - (if applicable) dimensions.
+     * and throws a semantic error if otherwise
+     */
+    void check_ArrayAssignmentCompatibility(ASR::expr_t* target, ASR::expr_t* value, const AST::Assignment_t &x) {
+        ASR::ttype_t *target_type = ASRUtils::type_get_past_allocatable(ASRUtils::expr_type(target));
+        ASR::ttype_t *value_type = ASRUtils::type_get_past_allocatable(ASRUtils::expr_type(value));
+        // we don't want to check by "check_equal_type" because we want to allow
+        // real :: x(4); x = [1, 2, 3, 4] to be a valid assignment (as RHS is "integer array")
+        if (target_type->type == ASR::ttypeType::Array && value_type->type == ASR::ttypeType::Array
+            && value->type != ASR::exprType::ArrayReshape) {
+            ASR::dimension_t* target_dims = nullptr;
+            ASR::dimension_t* value_dims = nullptr;
+            size_t target_rank = ASRUtils::extract_dimensions_from_ttype(target_type, target_dims);
+            size_t value_rank = ASRUtils::extract_dimensions_from_ttype(value_type, value_dims);
+            // ranks of LHS and RHS for an array assignment should
+            // be same (including allocatable arrays)
+            if (target_rank != value_rank) {
+                throw SemanticError("Incompatible ranks " + std::to_string(target_rank) +
+                   " and " + std::to_string(value_rank) + " in assignment",
+                  x.base.base.loc
+                );
+            } else if (ASRUtils::expr_type(target)->type == ASR::ttypeType::Array &&
+                       target_dims->m_length != nullptr &&
+                       !ASR::is_a<ASR::ArraySize_t>(*target_dims->m_length))
+            {
+                // if in any of the dimension, arrays have different size
+                // raise an error
+                for (size_t i = 0; i < target_rank; i++) {
+                    ASR::dimension_t dim_a = target_dims[i];
+                    ASR::dimension_t dim_b = value_dims[i];
+                    int dim_a_int {-1};
+                    int dim_b_int {-1};
+                    // 'm_length' isn't assigned for allocatable arrays
+                    // let them be valid for now atleast
+                    if (!(dim_a.m_length && dim_b.m_length)) {
+                        continue;
+                    }
+                    ASRUtils::extract_value(ASRUtils::expr_value(dim_a.m_length), dim_a_int);
+                    ASRUtils::extract_value(ASRUtils::expr_value(dim_b.m_length), dim_b_int);
+                    if (dim_a_int > 0 and dim_b_int > 0 && dim_a_int != dim_b_int) {
```

Comment:
To fix the Windows error:
```suggestion
                    if (dim_a_int > 0 && dim_b_int > 0 && dim_a_int != dim_b_int) {
```